File: _tbnf/src/apis.py
from __future__ import annotations
from typing import (Any, List, Tuple)
from ..fable_modules.fable_library.array import equals_with
from ..fable_modules.fable_library.list import (of_array, empty, FSharpList)
from ..fable_modules.fable_library.types import FSharpRef
from ..fable_modules.fable_library.util import equals
from .analysis import (build_analyzer as build_analyzer_1, Analyzer)
from .grammar import (Cell_1__ctor, Cell_1, position as position_1, expr, node as node_1, monot as monot_1, TTuple, TList, lexerule, definition, production, polyt, symbol)

def MK_Cell() -> Cell_1[Any]:
    return Cell_1__ctor()


def MK_position(line: int, col: int, filename: str) -> position_1:
    return position_1(line, col, filename)


def MK_EApp(f: expr, args: List[expr]) -> node_1:
    return node_1(0, f, of_array(args))


def MK_ETuple(args: List[expr]) -> node_1:
    return node_1(1, of_array(args))


def MK_EList(args: List[expr]) -> node_1:
    return node_1(2, of_array(args))


def MK_ELet(name: str, value: expr, body: expr) -> node_1:
    return node_1(4, name, value, body)


def MK_EFun(args: List[Tuple[str, monot_1]], body: expr) -> node_1:
    return node_1(5, of_array(args), body)


def MK_EVar(n: str) -> node_1:
    return node_1(6, n, FSharpRef(empty()))


def MK_ESlot(i: int) -> node_1:
    return node_1(7, i)


def MK_EInt(i: int) -> node_1:
    return node_1(8, i)


def MK_EStr(s: str) -> node_1:
    return node_1(9, s)


def MK_EFlt(s: float) -> node_1:
    return node_1(10, s)


def MK_EBool(b: bool) -> node_1:
    return node_1(11, b)


def MK_EField(value: expr, field: str) -> node_1:
    return node_1(3, value, field)


def MK_expr(node: node_1, pos: position_1) -> expr:
    return expr(node, pos, monot_1(1, "bot"))


def MK_TConst(n: str) -> monot_1:
    return monot_1(1, n)


def MK_TApp(f: monot_1, args: List[monot_1]) -> monot_1:
    return monot_1(2, f, of_array(args))


def MK_TFun(args: List[Tuple[str, monot_1]], ret: monot_1) -> monot_1:
    return monot_1(3, of_array(args), ret)


def MK_TVar(n: str) -> monot_1:
    return monot_1(4, n)


def MK_TTuple(xs: List[monot_1]) -> monot_1:
    if len(xs) == 1:
        return xs[0]
    
    else: 
        return TTuple(of_array(xs))
    


def MK_TList(a: monot_1) -> monot_1:
    return TList(a)


def MK_LStr(s: str) -> lexerule:
    return lexerule(3, s)


MK_LNumber : lexerule = lexerule(0)

MK_LWildcard : lexerule = lexerule(1)

def MK_LSeq(xs: List[lexerule]) -> lexerule:
    def arrow_350(x: lexerule, y: lexerule, xs: List[lexerule]=xs) -> bool:
        return equals(x, y)
    
    if len(xs) == 1 if (not equals_with(arrow_350, xs, None)) else (False):
        return xs[0]
    
    else: 
        return lexerule(2, of_array(xs))
    


def MK_LOr(xs: List[lexerule]) -> lexerule:
    def arrow_351(x: lexerule, y: lexerule, xs: List[lexerule]=xs) -> bool:
        return equals(x, y)
    
    if len(xs) == 1 if (not equals_with(arrow_351, xs, None)) else (False):
        return xs[0]
    
    else: 
        return lexerule(4, of_array(xs))
    


def MK_LNot(x: lexerule) -> lexerule:
    return lexerule(5, x)


def MK_LPlus(x: lexerule) -> lexerule:
    return lexerule(6, x)


def MK_LStar(x: lexerule) -> lexerule:
    return lexerule(7, x)


def MK_LOptional(x: lexerule) -> lexerule:
    return lexerule(8, x)


def MK_LGroup(x: lexerule) -> lexerule:
    return lexerule(9, x)


def MK_LRange(x: int, y: int) -> lexerule:
    return lexerule(10, x, y)


def MK_LRef(x: str) -> lexerule:
    return lexerule(11, x)


def MK_Defignore(ignore_list: List[str], pos: position_1) -> definition:
    return definition(6, {
        "ignoreList": of_array(ignore_list),
        "pos": pos
    })


def MK_Defmacro(lhs: str, parameters: List[str], define: List[Tuple[position_1, production]], pos: position_1) -> definition:
    def arrow_352(lhs: str=lhs, parameters: List[str]=parameters, define: List[Tuple[position_1, production]]=define, pos: position_1=pos) -> dict[str, Any]:
        parameters_1 : FSharpList[str] = of_array(parameters)
        return {
            "define": of_array(define),
            "lhs": lhs,
            "parameters": parameters_1,
            "pos": pos
        }
    
    return definition(0, arrow_352())


def MK_Defrule(lhs: str, define: List[Tuple[position_1, production]], pos: position_1) -> definition:
    return definition(1, {
        "define": of_array(define),
        "lhs": lhs,
        "pos": pos
    })


def MK_Deflexer(lhs: str, define: lexerule, pos: position_1) -> definition:
    return definition(2, {
        "define": define,
        "lhs": lhs,
        "pos": pos
    })


def MK_Declvar(ident: str, t: polyt, pos: position_1) -> definition:
    return definition(3, {
        "ident": ident,
        "pos": pos,
        "t": t
    })


def MK_Declctor(ident: str, t: monot_1, pos: position_1) -> definition:
    return definition(4, {
        "ident": ident,
        "pos": pos,
        "t": t
    })


def MK_Decltype(external: bool, has_fields: bool, ident: str, parameters: List[str], fields: List[Tuple[str, monot_1, position_1]], pos: position_1) -> definition:
    def arrow_353(external: bool=external, has_fields: bool=has_fields, ident: str=ident, parameters: List[str]=parameters, fields: List[Tuple[str, monot_1, position_1]]=fields, pos: position_1=pos) -> dict[str, Any]:
        parameters_1 : FSharpList[str] = of_array(parameters)
        return {
            "external": external,
            "fields": of_array(fields),
            "hasFields": has_fields,
            "ident": ident,
            "parameters": parameters_1,
            "pos": pos
        }
    
    return definition(5, arrow_353())


def MK_production(symbols: List[symbol], action: expr) -> production:
    return production(of_array(symbols), action)


def MK_Term(define: str, is_literal: bool) -> symbol:
    return symbol(0, define, is_literal)


def MK_Nonterm(define: str) -> symbol:
    return symbol(1, define)


def MK_Macrocall(n: str, syms: List[symbol], position: position_1) -> symbol:
    return symbol(2, n, of_array(syms), position)


def MK_Poly(bounds: List[str], monot: monot_1) -> polyt:
    return polyt(0, of_array(bounds), monot)


def MK_Mono(monot: monot_1) -> polyt:
    return polyt(1, monot)


def build_analyzer(stmts: List[definition]) -> Tuple[List[definition], Analyzer]:
    return build_analyzer_1(stmts)

File: tbnf-js/src/src/APIs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MK_Cell = MK_Cell;
exports.MK_Declctor = MK_Declctor;
exports.MK_Decltype = MK_Decltype;
exports.MK_Declvar = MK_Declvar;
exports.MK_Defignore = MK_Defignore;
exports.MK_Deflexer = MK_Deflexer;
exports.MK_Defmacro = MK_Defmacro;
exports.MK_Defrule = MK_Defrule;
exports.MK_EApp = MK_EApp;
exports.MK_EBool = MK_EBool;
exports.MK_EField = MK_EField;
exports.MK_EFlt = MK_EFlt;
exports.MK_EFun = MK_EFun;
exports.MK_EInt = MK_EInt;
exports.MK_ELet = MK_ELet;
exports.MK_EList = MK_EList;
exports.MK_ESlot = MK_ESlot;
exports.MK_EStr = MK_EStr;
exports.MK_ETuple = MK_ETuple;
exports.MK_EVar = MK_EVar;
exports.MK_LGroup = MK_LGroup;
exports.MK_LNot = MK_LNot;
exports.MK_LNumber = void 0;
exports.MK_LOptional = MK_LOptional;
exports.MK_LOr = MK_LOr;
exports.MK_LPlus = MK_LPlus;
exports.MK_LRange = MK_LRange;
exports.MK_LRef = MK_LRef;
exports.MK_LSeq = MK_LSeq;
exports.MK_LStar = MK_LStar;
exports.MK_LStr = MK_LStr;
exports.MK_LWildcard = void 0;
exports.MK_Macrocall = MK_Macrocall;
exports.MK_Mono = MK_Mono;
exports.MK_Nonterm = MK_Nonterm;
exports.MK_Poly = MK_Poly;
exports.MK_TApp = MK_TApp;
exports.MK_TConst = MK_TConst;
exports.MK_TFun = MK_TFun;
exports.MK_TList = MK_TList;
exports.MK_TTuple = MK_TTuple;
exports.MK_TVar = MK_TVar;
exports.MK_Term = MK_Term;
exports.MK_expr = MK_expr;
exports.MK_position = MK_position;
exports.MK_production = MK_production;
exports.build_analyzer = build_analyzer;

var _Grammar = require("./Grammar.js");

var _List = require("../fable_modules/fable-library.3.7.9/List.js");

var _Types = require("../fable_modules/fable-library.3.7.9/Types.js");

var _Array = require("../fable_modules/fable-library.3.7.9/Array.js");

var _Util = require("../fable_modules/fable-library.3.7.9/Util.js");

var _Analysis = require("./Analysis.js");

function MK_Cell() {
  return (0, _Grammar.Cell$1_$ctor)();
}

function MK_position(line, col, filename) {
  return new _Grammar.position(line, col, filename);
}

function MK_EApp(f, args) {
  return new _Grammar.node(0, f, (0, _List.ofArray)(args));
}

function MK_ETuple(args) {
  return new _Grammar.node(1, (0, _List.ofArray)(args));
}

function MK_EList(args) {
  return new _Grammar.node(2, (0, _List.ofArray)(args));
}

function MK_ELet(name, value, body) {
  return new _Grammar.node(4, name, value, body);
}

function MK_EFun(args, body) {
  return new _Grammar.node(5, (0, _List.ofArray)(args), body);
}

function MK_EVar(n) {
  return new _Grammar.node(6, n, new _Types.FSharpRef((0, _List.empty)()));
}

function MK_ESlot(i) {
  return new _Grammar.node(7, i);
}

function MK_EInt(i) {
  return new _Grammar.node(8, i);
}

function MK_EStr(s) {
  return new _Grammar.node(9, s);
}

function MK_EFlt(s) {
  return new _Grammar.node(10, s);
}

function MK_EBool(b) {
  return new _Grammar.node(11, b);
}

function MK_EField(value, field) {
  return new _Grammar.node(3, value, field);
}

function MK_expr(node, pos) {
  return new _Grammar.expr(node, pos, new _Grammar.monot(1, "bot"));
}

function MK_TConst(n) {
  return new _Grammar.monot(1, n);
}

function MK_TApp(f, args) {
  return new _Grammar.monot(2, f, (0, _List.ofArray)(args));
}

function MK_TFun(args, ret) {
  return new _Grammar.monot(3, (0, _List.ofArray)(args), ret);
}

function MK_TVar(n) {
  return new _Grammar.monot(4, n);
}

function MK_TTuple(xs) {
  if (xs.length === 1) {
    return xs[0];
  } else {
    return (0, _Grammar.TTuple)((0, _List.ofArray)(xs));
  }
}

function MK_TList(a) {
  return (0, _Grammar.TList)(a);
}

function MK_LStr(s) {
  return new _Grammar.lexerule(3, s);
}

const MK_LNumber = new _Grammar.lexerule(0);
exports.MK_LNumber = MK_LNumber;
const MK_LWildcard = new _Grammar.lexerule(1);
exports.MK_LWildcard = MK_LWildcard;

function MK_LSeq(xs) {
  if (!(0, _Array.equalsWith)(_Util.equals, xs, null) && xs.length === 1) {
    return xs[0];
  } else {
    return new _Grammar.lexerule(2, (0, _List.ofArray)(xs));
  }
}

function MK_LOr(xs) {
  if (!(0, _Array.equalsWith)(_Util.equals, xs, null) && xs.length === 1) {
    return xs[0];
  } else {
    return new _Grammar.lexerule(4, (0, _List.ofArray)(xs));
  }
}

function MK_LNot(x) {
  return new _Grammar.lexerule(5, x);
}

function MK_LPlus(x) {
  return new _Grammar.lexerule(6, x);
}

function MK_LStar(x) {
  return new _Grammar.lexerule(7, x);
}

function MK_LOptional(x) {
  return new _Grammar.lexerule(8, x);
}

function MK_LGroup(x) {
  return new _Grammar.lexerule(9, x);
}

function MK_LRange(x, y) {
  return new _Grammar.lexerule(10, x, y);
}

function MK_LRef(x) {
  return new _Grammar.lexerule(11, x);
}

function MK_Defignore(ignoreList, pos) {
  return new _Grammar.definition(6, {
    ignoreList: (0, _List.ofArray)(ignoreList),
    pos: pos
  });
}

function MK_Defmacro(lhs, parameters, define, pos) {
  let parameters_1;
  return new _Grammar.definition(0, (parameters_1 = (0, _List.ofArray)(parameters), {
    define: (0, _List.ofArray)(define),
    lhs: lhs,
    parameters: parameters_1,
    pos: pos
  }));
}

function MK_Defrule(lhs, define, pos) {
  return new _Grammar.definition(1, {
    define: (0, _List.ofArray)(define),
    lhs: lhs,
    pos: pos
  });
}

function MK_Deflexer(lhs, define, pos) {
  return new _Grammar.definition(2, {
    define: define,
    lhs: lhs,
    pos: pos
  });
}

function MK_Declvar(ident, t, pos) {
  return new _Grammar.definition(3, {
    ident: ident,
    pos: pos,
    t: t
  });
}

function MK_Declctor(ident, t, pos) {
  return new _Grammar.definition(4, {
    ident: ident,
    pos: pos,
    t: t
  });
}

function MK_Decltype(external, hasFields, ident, parameters, fields, pos) {
  let parameters_1;
  return new _Grammar.definition(5, (parameters_1 = (0, _List.ofArray)(parameters), {
    external: external,
    fields: (0, _List.ofArray)(fields),
    hasFields: hasFields,
    ident: ident,
    parameters: parameters_1,
    pos: pos
  }));
}

function MK_production(symbols, action) {
  return new _Grammar.production((0, _List.ofArray)(symbols), action);
}

function MK_Term(define, is_literal) {
  return new _Grammar.symbol(0, define, is_literal);
}

function MK_Nonterm(define) {
  return new _Grammar.symbol(1, define);
}

function MK_Macrocall(n, syms, position) {
  return new _Grammar.symbol(2, n, (0, _List.ofArray)(syms), position);
}

function MK_Poly(bounds, monot) {
  return new _Grammar.polyt(0, (0, _List.ofArray)(bounds), (0, _Grammar.processPolyType$0027)(bounds, monot));
}

function MK_Mono(monot) {
  return new _Grammar.polyt(1, monot);
}

function build_analyzer(stmts) {
  return (0, _Analysis.build_analyzer)(stmts);
}